### PR TITLE
feat(api)!: Web Socket State Machine 

### PR DESCRIPTION
**Pull request by Equartey**
*Description of changes:*
Refactor of web socket management into a state machine. Includes existing features while making functionality more robust and testable.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_core/lib/src/types/api/graphql/graphql_response.dart [hunk 1]
```diff
@@ -40,6 +40,6 @@ class GraphQLResponse<T> {
     );
   }
 
-  /// Whether errors are present in the response.
+  /// Returns true when errors are present and not empty.
```
Comment (ragingsquirrel3) [inline packages/amplify_core/lib/src/types/api/graphql/graphql_response.dart:43]: question: Was this change intended? maybe conflict error?

**Review by ragingsquirrel3 — APPROVED** [review]
approved w minor comment, great work again!

When you merge, could either be as rebase (keep the commits from this PR) or squashed. If you squash, I would copy the commit messages from this PR with individual PR links into the box for the detailed commit message so either way we see all the PRs in the commit history. 

Usually w a feature branch we merge w rebase to preserve history, but in this case I'm a little inclined to squash bc would be easier to revert if we had to.

File: packages/amplify_core/lib/src/types/api/graphql/graphql_response.dart (hunk 1 above)
Comment (dnys1) [reply packages/amplify_core/lib/src/types/api/graphql/graphql_response.dart:43]: Yes, please change back. Methods return, fields/getters hold information.

File: packages/api/amplify_api/lib/src/graphql/web_socket/services/web_socket_service.dart [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    } on Object catch (e, st) {
+      logger.error('Web socket error while initializing', e, st);
+      rethrow;
+    }
+  }
+
+  @override
+  Future<void> reconnect() {
+    // TODO(equartey): implement reconnect
+    throw UnimplementedError();
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/web_socket/services/web_socket_service.dart:115]: Needed still?

File: packages/api/amplify_api/lib/src/graphql/web_socket/services/web_socket_service.dart [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    );
+  }
+
+  void _send(WebSocketMessage message) {
+    try {
+      assert(sink != null, 'Sink must exist');
+      final msgJson = json.encode(message.toJson());
+      sink!.add(msgJson);
+    } on Exception catch (e) {
+      logger.warn(
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/web_socket/services/web_socket_service.dart:146]: What is the goal of this? Catching exceptions here won't do anything since `sink.add` will only ever throw an Error

File: packages/api/amplify_api/lib/src/graphql/web_socket/services/web_socket_service.dart [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+      case MessageType.stop:
+        return null;
+    }
+  }
+
+  @override
+  Future<void> close() async {
+    await sink?.close();
+    sink = null;
+  }
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/web_socket/services/web_socket_service.dart:227]: I would make this sync to avoid any weird race condition

```suggestion
  void close() {
    sink?.close();
    sink = null;
  }
```

File: packages/api/amplify_api/lib/src/graphql/web_socket/types/ws_subscriptions_event.dart [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:amplify_api/src/graphql/web_socket/blocs/web_socket_bloc.dart';
+import 'package:amplify_api/src/graphql/web_socket/types/web_socket_types.dart';
+
+/// Base Subscription event class
+abstract class WsSubscriptionEvent extends WebSocketEvent {
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/web_socket/types/ws_subscriptions_event.dart:19]: `SubscriptionEvent` is probably more fitting now

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/ws_subscriptions_bloc.dart [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+import 'package:amplify_api/src/graphql/web_socket/blocs/web_socket_bloc.dart';
+import 'package:amplify_api/src/graphql/web_socket/state/ws_subscriptions_state.dart';
+import 'package:amplify_api/src/graphql/web_socket/types/ws_subscriptions_event.dart';
+import 'package:amplify_core/amplify_core.dart';
+
+/// {@template amplify_api.ws_subscription_bloc}
+/// Internal state machine for subscriptions. Listens for [WsSubscriptionEvent]
+/// and maps them to appropriate state transitions.
+/// {@endtemplate}
+class WsSubscriptionBloc<T>
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/web_socket/blocs/ws_subscriptions_bloc.dart:27]: Can we call it `SubscriptionBloc`?

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/ws_subscriptions_bloc.dart [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    with AWSDebuggable, AmplifyLoggerMixin
+    implements Closeable {
+  /// {@macro api.ws_subscription_bloc}
+  /// takes in subscription events
+  /// transform/decode events
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/web_socket/blocs/ws_subscriptions_bloc.dart:32]:
```suggestion
  /// {@macro amplify_api.ws_subscription_bloc}
```

**Review by dnys1 — APPROVED** [review]
LGTM!

File: packages/api/amplify_api/lib/src/graphql/web_socket/services/web_socket_service.dart (hunk 3 above)
Comment (Equartey) [reply packages/api/amplify_api/lib/src/graphql/web_socket/services/web_socket_service.dart]: Discussed offline, removed in favor of more allow for more visibility of errors.
